### PR TITLE
chore(logging): Implement early logging

### PR DESCRIPTION
**Pull request by srebhan**
## Summary

This PR restructures the logging framework to implement "early" logging to collect log-messages that are issued before the actual logger is set-up. Currently, this is limited to non-fatal messages i.e. everything that survives loading and parsing of the config as well as configuring (but not initializing) the plugins.

## Checklist

- [x] No AI generated code was used in this PR

### CODE REVIEW

**Review by DStrand1 — APPROVED** [review]
LGTM!

**Review by powersj — COMMENTED** [review]
I think you mentioned this, but if I build and run by hand I see the loading config twice:

```sh
❯ ./telegraf --config config.toml
2024-07-18T16:19:12Z I! Loading config: config.toml
2024-07-18T16:19:12Z I! Loading config: config.toml
2024-07-18T16:19:12Z I! Starting Telegraf 1.32.0-ec5b839f brought to you by InfluxData the makers of InfluxDB
```

and with two configs:

```sh
❯ ./telegraf --config config.toml --config config.toml
2024-07-18T16:20:03Z I! Loading config: config.toml
2024-07-18T16:20:03Z I! Loading config: config.toml
2024-07-18T16:20:03Z W! Overlapping settings in multiple agent tables are not supported: may cause undefined behavior
2024-07-18T16:20:03Z I! Loading config: config.toml
2024-07-18T16:20:03Z I! Loading config: config.toml
2024-07-18T16:20:03Z W! Overlapping settings in multiple agent tables are not supported: may cause undefined behavior
```

The quiet and debug flags look fine.

**Review by powersj — COMMENTED** [review]
Got another one for you, it seems the debug agent output is now missing:

```toml
[agent]
  debug = true
  omit_hostname = true

[[inputs.mem]]

[[outputs.file]]

```

Specifically these lines:

```
2024-07-18T16:38:38Z D! [outputs.file]  Wrote batch of 1 metrics in 51.73µs
2024-07-18T16:38:38Z D! [outputs.file]  Buffer fullness: 0 / 10000 metrics
```

**Comment by srebhan** [comment]
@DStrand1 and @powersj please give this another review as the code structure changed _again_... Sorry!

File: plugins/common/shim/goshim.go [hunk 1]
```diff
@@ -77,6 +77,7 @@ func (s *Shim) watchForShutdown(cancel context.CancelFunc) {
 // Run the input plugins..
 func (s *Shim) Run(pollInterval time.Duration) error {
 	if s.Input != nil {
+		fmt.Printf("input logger: %+v (%T)\n", s.log, s.log)
```
Comment (DStrand1) [inline plugins/common/shim/goshim.go:80]: Looks line an unintentional printf?

**Review by DStrand1 — COMMENTED** [review]
Overall looks great! Just one small comment

File: plugins/common/shim/goshim.go (hunk 1 above)
Comment (powersj) [inline plugins/common/shim/goshim.go:80]:
```suggestion
```

**Pull request merged**